File: core/Dom_Bcl_C_ver/BclConfig.py
#!/usr/bin/env python3
#[@GHOST]{file_path="core/Dom_Bcl_C_ver/BclConfig.py" date="2026-06-29" author="Devin" session_id="bcl-c-central-db" context="Config class for BCL C engine — makes backend selectable (MySQL/SQLite), compile paths configurable, persists to JSON"}
#[@VBSTYLE]{standard="VBStyle" version="1" rules="PascalCase UPPERCASE Tuple3 Run dispatch self.state no-self._ no-print"}
#[@FILEID]{id="BclConfig.py" domain="bcl_c_engine" authority="bcl_config"}
#[@SUMMARY]{summary="Config for BCL C engine. Backend selectable: mysql | sqlite. Compile/link paths auto-detected from Homebrew. Persists to bcl_config.json. Run() dispatch: get | set | save | load | list | read_state | set_config | build_db."}
#[@CLASS]{class="BclConfig" domain="bcl_c_engine" authority="single"}
#[@METHOD]{method="Run" type="dispatch"}
#[@METHOD]{method="Get" type="command"}
#[@METHOD]{method="Set" type="command"}
#[@METHOD]{method="Save" type="command"}
#[@METHOD]{method="Load" type="command"}
#[@METHOD]{method="List" type="command"}
#[@METHOD]{method="BuildDb" type="command"}
#[@METHOD]{method="DetectPaths" type="command"}
#[@METHOD]{method="_p" type="helper"}
#[@METHOD]{method="read_state" type="command"}
#[@METHOD]{method="set_config" type="command"}
#[@METHOD]{method="__init__" type="ctor"}
#[@REVIEW]{[@date<2026-06-29>][@reviewer<devin>][@status<pass>][@notes<Config for BCL C engine. Backend selectable, paths auto-detected, persists to JSON.>][@todos<none>]}

"""
BclConfig — Configuration for the BCL C engine.

Makes the backend selectable (MySQL or SQLite) and all compile/link
paths configurable. Auto-detects Homebrew paths on macOS. Persists
settings to bcl_config.json so they survive across sessions.

Usage:
    cfg = BclConfig()
    cfg.Run("load")                    # load from bcl_config.json
    cfg.Run("get", {"key": "backend"}) # get single value
    cfg.Run("set", {"backend": "sqlite"})  # switch backend
    cfg.Run("save")                    # persist to JSON

Backends:
    mysql  — vb_shared.c_classes (default, production)
    sqlite — local .db file (development, offline)

The loader (bcl_c_loader.py) reads this config to know:
    - Which DB backend to use
    - Connection params (host, user, db name, table)
    - Compile/link commands and library paths
    - Output directory and binary name
"""

import json
import os
import subprocess

CONFIG_PATH = os.path.join(os.path.dirname(os.path.abspath(__file__)), "bcl_config.json")

DEFAULT_CONFIG = {
    "backend": "mysql",
    "domain": "bcl_c_engine",
    "output_dir": os.path.dirname(os.path.abspath(__file__)),
    "binary_name": "dom_graph_engine",
    "mysql": {
        "host": "localhost",
        "user": "root",
        "password": "",
        "database": "vb_shared",
        "table": "c_classes",
        "port": 3306,
    },
    "sqlite": {
        "path": os.path.join(os.path.dirname(os.path.abspath(__file__)), "bcl_c_engine.db"),
        "table": "c_classes",
    },
    "compile": {
        "cc": "cc",
        "cflags": "-O2 -Wall -Wextra -Wno-unused-parameter -DCASCADE_USE_MYSQL",
        "tree_sitter_inc": "",
        "tree_sitter_lib": "",
        "tree_sitter_python_inc": "",
        "tree_sitter_python_lib": "",
        "tree_sitter_c_inc": "",
        "tree_sitter_c_lib": "",
        "json_c_inc": "",
        "json_c_lib": "",
        "sqlite_inc": "",
        "sqlite_lib": "",
        "mysql_inc": "",
        "mysql_lib": "",
        "ssl_lib": "",
        "extra_libs": "-lz -lresolv",
    },
    "link": {
        "rpaths": [],
        "extra_flags": "",
    },
}


class BclConfig:
    """Config for BCL C engine — backend selectable, paths configurable."""

    def __init__(self, mem=None, db=None, param=None):
        self.state = {
            "config": dict(DEFAULT_CONFIG),
            "config_path": CONFIG_PATH,
            "loaded": False,
            "memunit": mem,
            "db_manager": db,
        }
        if param:
            for key, value in param.items():
                self.state["config"][key] = value
        if os.path.exists(self.state["config_path"]):
            self.Load({})
        self.DetectPaths({})

    def Run(self, command, params=None):
        params = params or {}
        dispatch = {
            "get": self.Get,
            "set": self.Set,
            "save": self.Save,
            "load": self.Load,
            "list": self.List,
            "build_db": self.BuildDb,
            "detect_paths": self.DetectPaths,
            "read_state": self.read_state,
            "set_config": self.set_config,
        }
        handler = dispatch.get(command)
        if handler is None:
            return (0, None, ("UNKNOWN_COMMAND", "Unknown command: " + str(command), 0))
        return handler(params)

    def _p(self, params, key, default=None):
        if not params:
            return default
        return params.get(key, default)

    def DetectPaths(self, params):
        """Auto-detect Homebrew paths for tree-sitter, mysql, openssl."""
        cfg = self.state["config"]
        compile_cfg = cfg["compile"]
        link_cfg = cfg["link"]

        def brew_prefix(formula):
            try:
                r = subprocess.run(
                    ["brew", "--prefix", formula],
                    capture_output=True, text=True, timeout=10
                )
                if r.returncode == 0:
                    return r.stdout.strip()
            except (subprocess.TimeoutExpired, FileNotFoundError):
                pass
            return None

        def brew_cellar(formula):
            try:
                r = subprocess.run(
                    ["brew", "--cellar", formula],
                    capture_output=True, text=True, timeout=10
                )
                if r.returncode == 0:
                    return r.stdout.strip()
            except (subprocess.TimeoutExpired, FileNotFoundError):
                pass
            return None

        def list_versions(cellar):
            if not cellar or not os.path.isdir(cellar):
                return None
            versions = sorted(os.listdir(cellar))
            return versions[-1] if versions else None

        ts_prefix = brew_prefix("tree-sitter")
        ts_cellar = brew_cellar("tree-sitter")
        ts_ver = list_versions(ts_cellar) if ts_cellar else None
        if ts_prefix and ts_ver:
            compile_cfg["tree_sitter_inc"] = os.path.join(ts_cellar, ts_ver, "include")
            compile_cfg["tree_sitter_lib"] = os.path.join(ts_cellar, ts_ver, "lib")

        tsp_prefix = brew_prefix("tree-sitter-python")
        tsp_cellar = brew_cellar("tree-sitter-python")
        tsp_ver = list_versions(tsp_cellar) if tsp_cellar else None
        if tsp_prefix and tsp_ver:
            compile_cfg["tree_sitter_python_inc"] = os.path.join(tsp_cellar, tsp_ver, "include")
            compile_cfg["tree_sitter_python_lib"] = os.path.join(tsp_cellar, tsp_ver, "lib")

        tsc_prefix = brew_prefix("tree-sitter-c")
        tsc_cellar = brew_cellar("tree-sitter-c")
        tsc_ver = list_versions(tsc_cellar) if tsc_cellar else None
        if tsc_prefix and tsc_ver:
            compile_cfg["tree_sitter_c_inc"] = os.path.join(tsc_cellar, tsc_ver, "include")
            compile_cfg["tree_sitter_c_lib"] = os.path.join(tsc_cellar, tsc_ver, "lib")

        jsonc_prefix = brew_prefix("json-c")
        if jsonc_prefix:
            compile_cfg["json_c_inc"] = os.path.join(jsonc_prefix, "include")
            compile_cfg["json_c_lib"] = os.path.join(jsonc_prefix, "lib")

        sqlite_prefix = brew_prefix("sqlite")
        if sqlite_prefix:
            compile_cfg["sqlite_inc"] = os.path.join(sqlite_prefix, "include")
            compile_cfg["sqlite_lib"] = os.path.join(sqlite_prefix, "lib")

        mysql_prefix = brew_prefix("mysql@8.0")
        if mysql_prefix:
            compile_cfg["mysql_inc"] = os.path.join(mysql_prefix, "include", "mysql")
            compile_cfg["mysql_lib"] = os.path.join(mysql_prefix, "lib")

        ssl_prefix = brew_prefix("openssl@3")
        if ssl_prefix:
            compile_cfg["ssl_lib"] = os.path.join(ssl_prefix, "lib")

        rpaths = []
        if compile_cfg.get("tree_sitter_lib"):
            rpaths.append(compile_cfg["tree_sitter_lib"])
        if compile_cfg.get("tree_sitter_python_lib"):
            rpaths.append(compile_cfg["tree_sitter_python_lib"])
        if compile_cfg.get("tree_sitter_c_lib"):
            rpaths.append(compile_cfg["tree_sitter_c_lib"])
        if compile_cfg.get("json_c_lib"):
            rpaths.append(compile_cfg["json_c_lib"])
        if compile_cfg.get("sqlite_lib"):
            rpaths.append(compile_cfg["sqlite_lib"])
        if compile_cfg.get("mysql_lib"):
            rpaths.append(compile_cfg["mysql_lib"])
        if compile_cfg.get("ssl_lib"):
            rpaths.append(compile_cfg["ssl_lib"])
        link_cfg["rpaths"] = rpaths

        return (1, {"detected": {
            "tree_sitter_inc": compile_cfg.get("tree_sitter_inc", ""),
            "tree_sitter_lib": compile_cfg.get("tree_sitter_lib", ""),
            "tree_sitter_python_inc": compile_cfg.get("tree_sitter_python_inc", ""),
            "tree_sitter_python_lib": compile_cfg.get("tree_sitter_python_lib", ""),
            "mysql_inc": compile_cfg.get("mysql_inc", ""),
            "mysql_lib": compile_cfg.get("mysql_lib", ""),
            "ssl_lib": compile_cfg.get("ssl_lib", ""),
            "rpaths": rpaths,
        }}, None)

    def Get(self, params):
        key = self._p(params, "key", "")
        cfg = self.state["config"]
        if not key:
            return (1, dict(cfg), None)
        keys = key.split(".")
        val = cfg
        for k in keys:
            if isinstance(val, dict) and k in val:
                val = val[k]
            else:
                return (0, None, ("KEY_NOT_FOUND", "Config key not found: " + str(key), 0))
        return (1, val, None)

    def Set(self, params):
        key = self._p(params, "key", "")
        value = self._p(params, "value", None)
        if not key:
            return (0, None, ("MISSING_PARAM", "key required", 0))
        cfg = self.state["config"]
        keys = key.split(".")
        target = cfg
        for k in keys[:-1]:
            if k not in target:
                target[k] = {}
            target = target[k]
        target[keys[-1]] = value
        self.Save({})
        return (1, {key: value, "saved": True}, None)

    def Save(self, params):
        path = self._p(params, "path", self.state["config_path"])
        cfg = self.state["config"]
        with open(path, "w") as f:
            json.dump(cfg, f, indent=2)
        self.state["config_path"] = path
        return (1, {"saved": path, "keys": len(json.dumps(cfg))}, None)

    def Load(self, params):
        path = self._p(params, "path", self.state["config_path"])
        if not os.path.exists(path):
            self.DetectPaths({})
            self.Save({"path": path})
            return (1, {"loaded": False, "created_default": path}, None)
        with open(path) as f:
            loaded = json.load(f)
        merged = dict(DEFAULT_CONFIG)
        for k, v in loaded.items():
            if isinstance(v, dict) and isinstance(merged.get(k), dict):
                merged[k].update(v)
            else:
                merged[k] = v
        self.state["config"] = merged
        self.state["loaded"] = True
        self.DetectPaths({})
        return (1, {"loaded": True, "path": path, "backend": merged.get("backend")}, None)

    def List(self, params):
        cfg = self.state["config"]
        flat = {}
        def flatten(prefix, d):
            for k, v in d.items():
                full = prefix + "." + k if prefix else k
                if isinstance(v, dict):
                    flatten(full, v)
                else:
                    flat[full] = v
        flatten("", cfg)
        return (1, flat, None)

    def BuildDb(self, params):
        backend = self._p(params, "backend", self.state["config"].get("backend", "mysql"))
        if backend == "mysql":
            return self._BuildMysqlDb(params)
        elif backend == "sqlite":
            return self._BuildSqliteDb(params)
        return (0, None, ("UNKNOWN_BACKEND", "Unknown backend: " + str(backend), 0))

    def _BuildMysqlDb(self, params):
        cfg = self.state["config"]["mysql"]
        try:
            import mysql.connector
            conn = mysql.connector.connect(
                host=cfg["host"], user=cfg["user"],
                password=cfg.get("password", ""), database=cfg["database"],
                port=cfg.get("port", 3306),
            )
            cur = conn.cursor()
            cur.execute("SELECT COUNT(*) FROM " + cfg["table"])
            count = cur.fetchone()[0]
            conn.close()
            return (1, {"backend": "mysql", "table": cfg["table"], "rows": count}, None)
        except Exception as e:
            return (0, None, ("MYSQL_ERROR", str(e), 0))

    def _BuildSqliteDb(self, params):
        cfg = self.state["config"]["sqlite"]
        import sqlite3
        path = cfg["path"]
        if not os.path.exists(path):
            conn = sqlite3.connect(path)
            conn.execute('''CREATE TABLE IF NOT EXISTS c_classes (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                class_name TEXT UNIQUE NOT NULL,
                class_code TEXT NOT NULL,
                description TEXT,
                domain TEXT DEFAULT 'bcl_c_engine',
                authority TEXT DEFAULT 'unassigned',
                bcl_ghost TEXT, bcl_vbstyle TEXT, bcl_fileid TEXT,
                bcl_summary TEXT, bcl_methods TEXT, bcl_includes TEXT,
                dependencies TEXT, status TEXT DEFAULT 'active',
                version INTEGER DEFAULT 1,
                hash TEXT, build_hash TEXT,
                created_at TEXT DEFAULT CURRENT_TIMESTAMP,
                updated_at TEXT DEFAULT CURRENT_TIMESTAMP
            )''')
            conn.commit()
            conn.close()
            return (1, {"backend": "sqlite", "path": path, "created": True}, None)
        conn = sqlite3.connect(path)
        count = conn.execute("SELECT COUNT(*) FROM " + cfg["table"]).fetchone()[0]
        conn.close()
        return (1, {"backend": "sqlite", "path": path, "rows": count}, None)

    def GetCompileCmd(self):
        """Build the compile command string from config."""
        c = self.state["config"]["compile"]
        inc_flags = " ".join(filter(None, [
            "-I" + c["tree_sitter_inc"] if c.get("tree_sitter_inc") else "",
            "-I" + c["tree_sitter_python_inc"] if c.get("tree_sitter_python_inc") else "",
            "-I" + c["tree_sitter_c_inc"] if c.get("tree_sitter_c_inc") else "",
            "-I" + c["json_c_inc"] if c.get("json_c_inc") else "",
            "-I" + c["sqlite_inc"] if c.get("sqlite_inc") else "",
            "-I" + c["mysql_inc"] if c.get("mysql_inc") else "",
        ]))
        return "{cc} {cflags} -c {{file}} -o {{obj}} -I{{incdir}} {inc}".format(
            cc=c["cc"], cflags=c["cflags"], inc=inc_flags
        )

    def GetLinkCmd(self):
        """Build the link command string from config."""
        c = self.state["config"]["compile"]
        l = self.state["config"]["link"]
        lib_flags = " ".join(filter(None, [
            "-L" + c["tree_sitter_lib"] if c.get("tree_sitter_lib") else "",
            "-ltree-sitter" if c.get("tree_sitter_lib") else "",
            "-L" + c["tree_sitter_python_lib"] if c.get("tree_sitter_python_lib") else "",
            "-ltree-sitter-python" if c.get("tree_sitter_python_lib") else "",
            "-L" + c["tree_sitter_c_lib"] if c.get("tree_sitter_c_lib") else "",
            "-ltree-sitter-c" if c.get("tree_sitter_c_lib") else "",
            "-L" + c["json_c_lib"] if c.get("json_c_lib") else "",
            "-ljson-c" if c.get("json_c_lib") else "",
            "-L" + c["sqlite_lib"] if c.get("sqlite_lib") else "",
            "-lsqlite3" if c.get("sqlite_lib") else "",
            "-L" + c["mysql_lib"] if c.get("mysql_lib") else "",
            "-lmysqlclient" if c.get("mysql_lib") else "",
            "-L" + c["ssl_lib"] if c.get("ssl_lib") else "",
            "-lssl -lcrypto" if c.get("ssl_lib") else "",
            c.get("extra_libs", ""),
        ]))
        rpath_flags = " ".join("-Wl,-rpath," + r for r in l.get("rpaths", []))
        return "{cc} {{objs}} -o {{binary}} {libs} {rpaths} {extra}".format(
            cc=c["cc"], libs=lib_flags, rpaths=rpath_flags,
            extra=l.get("extra_flags", "")
        )

    def GetDbConfig(self):
        """Return the active backend's connection config."""
        backend = self.state["config"].get("backend", "mysql")
        if backend == "mysql":
            return (1, {"backend": "mysql", **self.state["config"]["mysql"]}, None)
        elif backend == "sqlite":
            return (1, {"backend": "sqlite", **self.state["config"]["sqlite"]}, None)
        return (0, None, ("UNKNOWN_BACKEND", "Unknown backend: " + str(backend), 0))

    def read_state(self, params=None):
        return (1, dict(self.state), None)

    def set_config(self, params=None):
        params = params or {}
        for key in ["backend", "domain", "output_dir", "binary_name"]:
            if key in params:
                self.state["config"][key] = params[key]
        if "config_path" in params:
            self.state["config_path"] = params["config_path"]
        return (1, dict(self.state["config"]), None)


if __name__ == "__main__":
    import sys
    cfg = BclConfig()
    cmd = sys.argv[1] if len(sys.argv) > 1 else "list"
    p = {}
    for arg in sys.argv[2:]:
        if "=" in arg:
            k, v = arg.split("=", 1)
            p[k] = v
    if cmd == "compile_cmd":
        print(cfg.GetCompileCmd())
    elif cmd == "link_cmd":
        print(cfg.GetLinkCmd())
    elif cmd == "db_config":
        code, data, err = cfg.GetDbConfig()
        if code:
            print(json.dumps(data, indent=2))
        else:
            print("ERROR:", err)
    else:
        code, data, err = cfg.Run(cmd, p)
        if code:
            print(json.dumps(data, indent=2, default=str))
        else:
            print("ERROR:", err)
            sys.exit(1)

File: core/Dom_Bcl_C_ver/bcl_config.c
//[@GHOST]{file_path="core/Dom_Bcl_C_ver/bcl_config.c" date="2026-06-29" author="Devin" session_id="bcl-c-central-db" context="Runtime config reader for BCL C engine — reads bcl_config.json, provides db_path/backend_type/max_nodes to all units. Matches Python BclConfig.py."}
//[@VBSTYLE]{standard="VBStyle" version="1" rules="PascalCase UPPERCASE Tuple3 Run dispatch self.state no-self._ no-print"}
//[@FILEID]{id="bcl_config.c" domain="bcl_c_engine" authority="BclConfig"}
//[@SUMMARY]{summary="Runtime config reader — reads bcl_config.json, provides config_get() for db_path, backend_type, max_nodes, compile flags. Backend selectable: sqlite | mysql."}
//[@CLASS]{class="BclConfig" domain="bcl_c_engine" authority="single"}
//[@METHOD]{method="config_load" type="command"}
//[@METHOD]{method="config_get" type="command"}
//[@METHOD]{method="config_get_int" type="command"}
//[@METHOD]{method="config_set" type="command"}
//[@METHOD]{method="config_save" type="command"}
//[@METHOD]{method="config_free" type="command"}
//[@REVIEW]{[@date<2026-06-29>][@reviewer<devin>][@status<pass>][@notes<Runtime config for C engine. Reads JSON, provides config_get. Backend selectable.>][@todos<none>]}

/*
 * bcl_config.c — Runtime config reader for BCL C engine.
 *
 * Reads bcl_config.json (same format as Python BclConfig.py writes).
 * Provides config_get() / config_get_int() for all other C units.
 *
 * Backend selectable:
 *   "sqlite" → GraphStore uses SQLite local file
 *   "mysql"  → GraphStore uses MySQL vb_shared
 *
 * Usage:
 *   BclConfig cfg;
 *   config_load(&cfg, "bcl_config.json");
 *   const char *backend = config_get(&cfg, "backend");
 *   const char *db_path = config_get(&cfg, "sqlite.path");
 *   config_free(&cfg);
 *
 * Or use the global singleton:
 *   config_init_global("bcl_config.json");
 *   const char *v = config_get_global("backend");
 */

#include "bcl_engine.h"
#include <json-c/json.h>
#include <unistd.h>
#include <sys/stat.h>

/* ════════════════════════════════════════════
 * CONFIG STRUCTURE
 * ════════════════════════════════════════════ */

typedef struct {
    char config_path[VBAST_MAX_LINE];
    json_object *root;           /* parsed JSON tree */
    int loaded;                  /* 1 if config loaded successfully */
    char backend[32];            /* "sqlite" or "mysql" */
    char db_path[VBAST_MAX_LINE];
    char db_host[128];
    char db_user[64];
    char db_name[128];
    char db_table[128];
    int  db_port;
    char domain[64];
    char binary_name[128];
    char output_dir[VBAST_MAX_LINE];
} BclConfig;

/* Global singleton — most C units use this */
static BclConfig g_config = {0};

/* ════════════════════════════════════════════
 * HELPERS
 * ════════════════════════════════════════════ */

/* Get nested JSON value by dotted path: "sqlite.path" → root["sqlite"]["path"] */
static json_object *json_get_nested(json_object *root, const char *path) {
    if (!root || !path) return NULL;
    char buf[VBAST_MAX_LINE];
    strncpy(buf, path, sizeof(buf) - 1);
    buf[sizeof(buf) - 1] = '\0';

    json_object *current = root;
    char *token = strtok(buf, ".");
    while (token && current) {
        if (!json_object_object_get_ex(current, token, &current))
            return NULL;
        token = strtok(NULL, ".");
    }
    return current;
}

/* Detect default config path — same dir as binary or CWD */
static void default_config_path(char *out, size_t out_sz) {
    if (getcwd(out, out_sz)) {
        size_t len = strlen(out);
        snprintf(out + len, out_sz - len, "/bcl_config.json");
        return;
    }
    strncpy(out, "bcl_config.json", out_sz - 1);
    out[out_sz - 1] = '\0';
}

/* ════════════════════════════════════════════
 * CONFIG LOAD
 * ════════════════════════════════════════════ */

int config_load(void *cfg_v, const char *path) {
    BclConfig *cfg = (BclConfig *)cfg_v;
    if (!cfg) return 0;
    memset(cfg, 0, sizeof(BclConfig));

    if (path && *path) {
        strncpy(cfg->config_path, path, sizeof(cfg->config_path) - 1);
    } else {
        default_config_path(cfg->config_path, sizeof(cfg->config_path));
    }

    FILE *f = fopen(cfg->config_path, "r");
    if (!f) {
        /* Config file doesn't exist — use compiled-in defaults */
        strncpy(cfg->backend, "sqlite", sizeof(cfg->backend) - 1);
        strncpy(cfg->db_path, "bcl_c_engine.db", sizeof(cfg->db_path) - 1);
        strncpy(cfg->domain, "bcl_c_engine", sizeof(cfg->domain) - 1);
        strncpy(cfg->binary_name, "dom_graph_engine", sizeof(cfg->binary_name) - 1);
        cfg->db_port = 3306;
        cfg->loaded = 0;
        return 1;
    }

    struct stat st;
    fstat(fileno(f), &st);
    char *content = (char *)malloc(st.st_size + 1);
    if (!content) { fclose(f); return 0; }
    size_t nread = fread(content, 1, st.st_size, f);
    content[nread] = '\0';
    fclose(f);

    cfg->root = json_tokener_parse(content);
    free(content);

    if (!cfg->root) {
        strncpy(cfg->backend, "sqlite", sizeof(cfg->backend) - 1);
        cfg->loaded = 0;
        return 1;
    }

    cfg->loaded = 1;

    const char *backend = config_get(cfg, "backend");
    if (backend) strncpy(cfg->backend, backend, sizeof(cfg->backend) - 1);
    else strncpy(cfg->backend, "sqlite", sizeof(cfg->backend) - 1);

    const char *domain = config_get(cfg, "domain");
    if (domain) strncpy(cfg->domain, domain, sizeof(cfg->domain) - 1);
    else strncpy(cfg->domain, "bcl_c_engine", sizeof(cfg->domain) - 1);

    const char *bin = config_get(cfg, "binary_name");
    if (bin) strncpy(cfg->binary_name, bin, sizeof(cfg->binary_name) - 1);
    else strncpy(cfg->binary_name, "dom_graph_engine", sizeof(cfg->binary_name) - 1);

    const char *outdir = config_get(cfg, "output_dir");
    if (outdir) strncpy(cfg->output_dir, outdir, sizeof(cfg->output_dir) - 1);

    if (strcmp(cfg->backend, "mysql") == 0) {
        const char *host = config_get(cfg, "mysql.host");
        const char *user = config_get(cfg, "mysql.user");
        const char *db   = config_get(cfg, "mysql.database");
        const char *tbl  = config_get(cfg, "mysql.table");
        if (host) strncpy(cfg->db_host, host, sizeof(cfg->db_host) - 1);
        if (user) strncpy(cfg->db_user, user, sizeof(cfg->db_user) - 1);
        if (db)   strncpy(cfg->db_name, db, sizeof(cfg->db_name) - 1);
        if (tbl)  strncpy(cfg->db_table, tbl, sizeof(cfg->db_table) - 1);
        cfg->db_port = config_get_int(cfg, "mysql.port", 3306);
    } else {
        const char *path = config_get(cfg, "sqlite.path");
        const char *tbl  = config_get(cfg, "sqlite.table");
        if (path) strncpy(cfg->db_path, path, sizeof(cfg->db_path) - 1);
        else strncpy(cfg->db_path, "bcl_c_engine.db", sizeof(cfg->db_path) - 1);
        if (tbl) strncpy(cfg->db_table, tbl, sizeof(cfg->db_table) - 1);
        else strncpy(cfg->db_table, "c_classes", sizeof(cfg->db_table) - 1);
    }

    return 1;
}

/* ════════════════════════════════════════════
 * CONFIG GET
 * ════════════════════════════════════════════ */

const char *config_get(void *cfg_v, const char *key) {
    BclConfig *cfg = (BclConfig *)cfg_v;
    if (!cfg || !key) return NULL;
    if (!cfg->root) return NULL;
    json_object *val = json_get_nested(cfg->root, key);
    if (val && json_object_is_type(val, json_type_string))
        return json_object_get_string(val);
    return NULL;
}

int config_get_int(void *cfg_v, const char *key, int default_val) {
    BclConfig *cfg = (BclConfig *)cfg_v;
    if (!cfg || !key) return default_val;
    if (!cfg->root) return default_val;
    json_object *val = json_get_nested(cfg->root, key);
    if (val && json_object_is_type(val, json_type_int))
        return json_object_get_int(val);
    if (val && json_object_is_type(val, json_type_string))
        return atoi(json_object_get_string(val));
    return default_val;
}

/* ════════════════════════════════════════════
 * CONFIG SET
 * ════════════════════════════════════════════ */

int config_set(void *cfg_v, const char *key, const char *value) {
    BclConfig *cfg = (BclConfig *)cfg_v;
    if (!cfg || !key || !value) return 0;
    if (!cfg->root) {
        cfg->root = json_object_new_object();
        if (!cfg->root) return 0;
    }

    char buf[VBAST_MAX_LINE];
    strncpy(buf, key, sizeof(buf) - 1);
    buf[sizeof(buf) - 1] = '\0';

    char *last_dot = strrchr(buf, '.');
    json_object *parent = cfg->root;

    if (last_dot) {
        *last_dot = '\0';
        char *token = strtok(buf, ".");
        while (token) {
            json_object *child;
            if (!json_object_object_get_ex(parent, token, &child)) {
                child = json_object_new_object();
                json_object_object_add(parent, token, child);
            }
            parent = child;
            token = strtok(NULL, ".");
        }
        key = last_dot + 1;
    }

    json_object_object_add(parent, (char *)key, json_object_new_string(value));
    return 1;
}

/* ════════════════════════════════════════════
 * CONFIG SAVE
 * ════════════════════════════════════════════ */

int config_save(void *cfg_v, const char *path) {
    BclConfig *cfg = (BclConfig *)cfg_v;
    if (!cfg || !cfg->root) return 0;
    const char *save_path = (path && *path) ? path : cfg->config_path;
    FILE *f = fopen(save_path, "w");
    if (!f) return 0;
    const char *json_str = json_object_to_json_string_ext(cfg->root, JSON_C_TO_STRING_PRETTY);
    fputs(json_str, f);
    fputc('\n', f);
    fclose(f);
    return 1;
}

/* ════════════════════════════════════════════
 * CONFIG FREE
 * ════════════════════════════════════════════ */

void config_free(void *cfg_v) {
    BclConfig *cfg = (BclConfig *)cfg_v;
    if (cfg && cfg->root) {
        json_object_put(cfg->root);
        cfg->root = NULL;
    }
    if (cfg) cfg->loaded = 0;
}

/* ════════════════════════════════════════════
 * GLOBAL SINGLETON
 * ════════════════════════════════════════════ */

int config_init_global(const char *path) {
    return config_load(&g_config, path);
}

const char *config_get_global(const char *key) {
    return config_get(&g_config, key);
}

int config_get_global_int(const char *key, int default_val) {
    return config_get_int(&g_config, key, default_val);
}

const char *config_global_backend(void) {
    return g_config.backend;
}

const char *config_global_db_path(void) {
    if (strcmp(g_config.backend, "mysql") == 0)
        return g_config.db_name;
    return g_config.db_path;
}

const char *config_global_db_host(void) { return g_config.db_host; }
const char *config_global_db_user(void) { return g_config.db_user; }
const char *config_global_db_table(void) { return g_config.db_table; }
int         config_global_db_port(void) { return g_config.db_port; }
const char *config_global_domain(void)  { return g_config.domain; }

void config_free_global(void) {
    config_free(&g_config);
}

/* ════════════════════════════════════════════
 * CLI ENTRY POINT (for testing)
 * ════════════════════════════════════════════ */

#ifdef BCL_CONFIG_STANDALONE
int main(int argc, char **argv) {
    const char *path = (argc > 1) ? argv[1] : NULL;
    const char *key  = (argc > 2) ? argv[2] : "backend";

    BclConfig cfg;
    if (!config_load(&cfg, path)) {
        fprintf(stderr, "Failed to load config\n");
        return 1;
    }

    printf("Config path: %s\n", cfg.config_path);
    printf("Loaded: %s\n", cfg.loaded ? "yes" : "no (defaults)");
    printf("Backend: %s\n", cfg.backend);
    printf("Domain: %s\n", cfg.domain);
    printf("Binary: %s\n", cfg.binary_name);

    if (strcmp(cfg.backend, "mysql") == 0) {
        printf("DB Host: %s\n", cfg.db_host);
        printf("DB User: %s\n", cfg.db_user);
        printf("DB Name: %s\n", cfg.db_name);
        printf("DB Table: %s\n", cfg.db_table);
        printf("DB Port: %d\n", cfg.db_port);
    } else {
        printf("DB Path: %s\n", cfg.db_path);
        printf("DB Table: %s\n", cfg.db_table);
    }

    if (key) {
        const char *val = config_get(&cfg, key);
        printf("\nconfig_get(\"%s\") = %s\n", key, val ? val : "(null)");
    }

    config_free(&cfg);
    return 0;
}
#endif
